docs/codeql/codeql-language-guides/customizing-library-models-for-java.rst

```diff
@@ -0,0 +1,439 @@
+- The third value **False** is a flag that indicates whether or not the summary also applies to all overrides of the method.
+- The fourth value **concat** is the method name.
+- The fifth value **(String)** is the method input type signature.
+
+The sixth value is only relevant internally and can be omitted in most use cases.
+The remaining values are used to define the **access path**, the **kind**, and the **provenance** (origin) of the summary.
+
+- The seventh value is the access path to the input (where data flows from). **Argument[this]** is the access path to the qualifier (**s1** in the example) and **Argument[0]** is the access path to the first argument (**s2** in the example).
+- The eighth value **ReturnValue** is the access path to the output (where data flows to), in this case **ReturnValue**, which means that the input flows to the return value.
+- The ninth value **taint** is the kind of the flow. **taint** means that taint is propagated through the call.
+- The tenth value **manual** is the provenance of the summary, which is used to identify the origin of the summary.
+
+Example: Add flow through the **map** method.
+---------------------------------------------
+In this example, we will see a more complex example of modeling flow through a method.
+This pattern shows how to model flow through higher order methods and collection types.
+Please note that the flow through the **map** method is already added to the CodeQL Java analysis.
+
+.. code-block:: java
+
+   public static void taintflow(Stream<String> s) {
+     Stream<String> l = s.map(e -> e.concat("\n"));
+     ...
+   }
+
+This can be achieved by adding the following to a data extension file:
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/java-all
+         extensible: summaryModel
+       data:
+         - ["java.util.stream", "Stream", True, "map", "(Function)", "", "Argument[this].Element", "Argument[0].Parameter[0]", "value", "manual"]
+         - ["java.util.stream", "Stream", True, "map", "(Function)", "", "Argument[0].ReturnValue", "ReturnValue.Element", "value", "manual"]
+
+Reasoning:
+
+Since we are adding flow through a method, we need to add tuples to the **summaryModel** extensible predicate.
+Each tuple defines part of the flow that comprises the total flow through the **map** method.
+The first five values identify the callable (in this case a method) to be modeled as a summary.
+These are the same for both of the rows above as we are adding two summaries for the same method.
+
+- The first value **java.util.stream** is the package name.
+- The second value **Stream** is the class (type) name.
+- The third value **True** is a flag that indicates whether or not the summary also applies to all overrides of the method.
+- The fourth value **map** is the method name.
+- The fifth value **Function** is the method input type signature.
+
+The sixth value is only relevant internally and can be omitted in most use cases.
```

Comment:
Same comment as above but ... It might be worth clarifying what you mean by "only relevant internally" but I may be betraying my layman status here so feel free to ignore this